src/blas/backends/cublas/cublas_scope_handle.cpp

```diff
@@ -17,126 +17,28 @@
 *
 **************************************************************************/
 #include "cublas_scope_handle.hpp"
-#if __has_include(<sycl/detail/common.hpp>)
-#include <sycl/detail/common.hpp>
-#else
-#include <CL/sycl/detail/common.hpp>
-#endif
 
 namespace oneapi {
 namespace mkl {
 namespace blas {
 namespace cublas {
 
-/**
- * Inserts a new element in the map if its key is unique. This new element
- * is constructed in place using args as the arguments for the construction
- * of a value_type (which is an object of a pair type). The insertion only
- * takes place if no other element in the container has a key equivalent to
- * the one being emplaced (keys in a map container are unique).
- */
-#ifdef ONEMKL_PI_INTERFACE_REMOVED
-thread_local cublas_handle<ur_context_handle_t> CublasScopedContextHandler::handle_helper =
-    cublas_handle<ur_context_handle_t>{};
-#else
-thread_local cublas_handle<pi_context> CublasScopedContextHandler::handle_helper =
-    cublas_handle<pi_context>{};
-#endif
-
-CublasScopedContextHandler::CublasScopedContextHandler(sycl::queue queue, sycl::interop_handle& ih)
+CublasScopedContextHandler::CublasScopedContextHandler(sycl::interop_handle& ih)
         : ih(ih),
-          needToRecover_(false) {
-    placedContext_ = new sycl::context(queue.get_context());
-    auto cudaDevice = ih.get_native_device<sycl::backend::ext_oneapi_cuda>();
-    CUresult err;
-    CUcontext desired;
-    CUDA_ERROR_FUNC(cuCtxGetCurrent, err, &original_);
-    CUDA_ERROR_FUNC(cuDevicePrimaryCtxRetain, err, &desired, cudaDevice);
-    if (original_ != desired) {
-        // Sets the desired context as the active one for the thread
-        CUDA_ERROR_FUNC(cuCtxSetCurrent, err, desired);
-        // No context is installed and the suggested context is primary
-        // This is the most common case. We can activate the context in the
-        // thread and leave it there until all the PI context referring to the
-        // same underlying CUDA primary context are destroyed. This emulates
-        // the behaviour of the CUDA runtime api, and avoids costly context
-        // switches. No action is required on this side of the if.
-        needToRecover_ = !(original_ == nullptr);
-    }
-}
-
-CublasScopedContextHandler::~CublasScopedContextHandler() noexcept(false) {
-    if (needToRecover_) {
-        CUresult err;
-        CUDA_ERROR_FUNC(cuCtxSetCurrent, err, original_);
-    }
-    delete placedContext_;
-}
-
-void ContextCallback(void* userData) {
-    auto* ptr = static_cast<std::atomic<cublasHandle_t>*>(userData);
-    if (!ptr) {
-        return;
-    }
-    auto handle = ptr->exchange(nullptr);
-    if (handle != nullptr) {
-        cublasStatus_t err1;
-        CUBLAS_ERROR_FUNC(cublasDestroy, err1, handle);
```

Comment:
It seems we never destroy the cublas handle anymore. I get that the existing solution relies on `sycl::detail::pi::contextSetExtendedDeleter` which is not ideal, is there no other way we could destroy the cublas handle when the SYCL queue is destroyed? This seems like a useful feature.

Reply:
Right I forgot the handle could also be destroyed when destroying `cublas_handle`.
I'm not sure it is possible to destroy it at the end of the program with the current solution. You have to be very careful about the fact that `host_task` are used which spawn their own threads. This means every `host_task` creates its own `CublasScopedContextHandler`, have their own copy of `cublas_handle` as it is `thread_local` in the develop branch.
I am worried that your suggested solution would still create and destroy the cublas handle in every host task. I don't know how this could be improved, I am just very wary of any change in the scope handle!

Reply:
Yes, it's been quite a ride trying to untangle all of this so far! :wink: 

I think if that's the case with the host task, we just need to make the member `cublasHandle_t` thread_local and shared ptr and add a destructor to it which destroys it at the end of the program. 

I'm actually off sick today so I will think about it more and get back to you! Thanks for the review tho, it's much appreciated.